### ISSUE TITLE
Getting infinite call loop related to fish_user_key_bindings

### ISSUE THREAD

**Issue by EdmundsEcho** [fish-shell/fish-shell#7557]
The error occurs when I source my `config.fish` file.  It has been working for a couple of years.  "it" in particular, meaning the config in general, and the use of the function that is at the heart of the error message: `fish_user_key_bindings`.  Specifically, here is the settings:
```fishshell
# config.fish - unchanged
set -g fish_key_bindings my_vi_bindings
```

```fishshell
# functions/my_vi_bindings - unchanged
# Key bindings for the vi-mode plugin
function my_vi_bindings
  fish_vi_key_bindings
  bind -M insert -m default df backward-char force-repaint
end
```

```fishshell
# Not me! :)
# This file contains fish universal variable definitions.
# VERSION: 3.0
SETUVAR --export PYENV_ROOT:/Users/edmund/\x2epyenv
SETUVAR ZO_CMD:zo
SETUVAR Z_CMD:z
SETUVAR Z_DATA:/Users/edmund/\x2elocal/share/z/data
SETUVAR Z_DATA_DIR:/Users/edmund/\x2elocal/share/z
SETUVAR Z_EXCLUDE:/Users/edmund
SETUVAR __fish_init_2_39_8:\x1d
SETUVAR __fish_init_2_3_0:\x1d
SETUVAR __fish_init_3_x:\x1d
SETUVAR __fish_initialized:3100
SETUVAR fish_color_autosuggestion:BD93F9
SETUVAR fish_color_cancel:normal
SETUVAR fish_color_command:8BE9FD
SETUVAR fish_color_comment:6272A4
SETUVAR fish_color_cwd:008000
SETUVAR fish_color_cwd_root:800000
SETUVAR fish_color_end:50FA7B
SETUVAR fish_color_error:FFB86C
SETUVAR fish_color_escape:00a6b2
SETUVAR fish_color_history_current:normal
SETUVAR fish_color_host:normal
SETUVAR fish_color_host_remote:yellow
SETUVAR fish_color_match:normal
SETUVAR fish_color_normal:normal
SETUVAR fish_color_operator:00a6b2
SETUVAR fish_color_param:00afff
SETUVAR fish_color_quote:F1FA8C
SETUVAR fish_color_redirection:8BE9FD
SETUVAR fish_color_search_match:ffff00
SETUVAR fish_color_selection:c0c0c0
SETUVAR fish_color_status:red
SETUVAR fish_color_user:00ff00
SETUVAR fish_color_valid_path:normal
SETUVAR fish_greeting:Welcome\x20to\x20fish\x2c\x20the\x20friendly\x20interactive\x20shell

# 🚨 The following line looks suspicious
SETUVAR fish_key_bindings:fish_vi_key_bindings

SETUVAR fish_pager_color_completion:normal
SETUVAR fish_pager_color_description:B3A06D\x1eyellow
SETUVAR fish_pager_color_prefix:white\x1e\x2d\x2dbold\x1e\x2d\x2dunderline
SETUVAR fish_pager_color_progress:brwhite\x1e\x2d\x2dbackground\x3dcyan
```

```fishshell
# the error
fish: The function call stack limit has been exceeded. Do you have an accidental infinite loop?
                for bindings in $__fundle_binding_paths
                        source $bindings
                end
                if functions -q __fish_user_key_bindings
                        __fish_user_key_bindings
                end
                ^
in function '__fish_user_key_bindings'
...
...
in function '__fish_user_key_bindings'
        called on line 285 of file ~/.config/fish/config.fish
in function 'fish_user_key_bindings'
in function '__fish_reload_key_bindings' with arguments 'VARIABLE SET fish_key_bindings'
        called on line 1 of file ~/.config/fish/config.fish
in event handler: handler for variable 'fish_key_bindings'
        called on line 159 of file ~/.config/fish/config.fish
read config.fish

```

I'm not sure of the cause but it renders the config unuseful.  The trigger for all of this was my adding then removing an alias definition.  The "crash" occurred once I removed the alias.  In my experience, updating an alias triggers auto-magic function updates.

The "fix": remove the setting in my config, witness the removal of the offending `SETUVAR fish_key_bindings:fish_vi_key_bindings`. In this state, I can load the config file without seeing the error.  However, in a fresh terminal, I lose the vi-like functionality.  I can restore the feature by re-activating the setting in my config, watch it generate the error, but none-the-less use the terminal.

Thank you for any pointers on what might be going on here.

\- E

**Comment by faho** [comment]
First of all: You're overthinking this.

Just put

```fish
bind -M insert -m default df backward-char force-repaint
```

into config.fish, set fish_key_bindings to fish_vi_key_bindings and you're done. There's no need to define your own binding function just to add a few bindings. If you want to organize your own bindings, put them into a function called "fish_user_key_bindings" - fish automatically calls that after setting up the bindings.

-------

What seems to be happening here is that the function `fish_vi_key_bindings` will, when it's called, check if $fish_key_bindings is set to "fish_vi_key_bindings", and if not, it will set it and return. Setting the variable triggers a variable handler that then calls the function mentioned in the variable.

This is so you can change keybindings either via `set fish_key_bindings` or by calling the function - it's a bit rickety, but backwards compatibility and such. (tbh we should probably just drop support for one or the other)

So you set $fish_key_bindings to your function, which causes your function to be called, which calls fish_vi_key_bindings, which sets $fish_key_bindings to "fish_vi_key_bindings", which calls fish_vi_key_bindings, and since the variable is set to "fish_vi_key_bindings" this time it runs to completion, and you have $fish_key_bindings now set to fish_vi_key_bindings.

Only... then it goes on to call `fish_user_key_bindings` to apply your custom bindings, and fundle handles that specially by providing its own function called `fish_user_key_bindings` that does two things:

- `source` all the files in a variable called `__fundle_binding_paths` - which defaults to all files called "key_bindings.fish" in any plugins you're using

- call your actual `fish_user_key_bindings` which it moved out of the way

And, well, probably some plugin here modifies $fish_key_bindings, causing the entire dance to repeat. (and there are more plugins that do that than you'd assume - we've had *prompts* declare that they "require" vi-mode and enable it).

**Comment by EdmundsEcho** [comment]
Thank you! I am often guilty of overthinking it.  Thank goodness for the advice of others to keep it real.  

I just looked up the docs for `bind` - who knew!  It's one thing to move from word-processor to word-processor, it's a whole other capacity to know that you can "copy/paste".  

Inspired by your thoughtful post, and my ever trying to understand how "I missed it", and why, this post morphed into "this person's experience with fishshell".  To the degree describing a user experience is critical to accelerating useful innovations,  I hope this is useful.

So, I bring up the "copy/paste" analogy because I've been using fishshell for over 2-3 years now.  The different syntax between bash (and the like) and fishshell is clear and present; I can anticipate what I need to learn when I bump into script snippets accordingly.  

But key bindings? I put that in one of a couple of buckets: ApplesScript + system prefs (perfectly good tech, but not for me), shell aliases, and fishshell functions.  Finally, I've been using Unix off and on for decades; 100% of the time the day Apple moved to a Unix-based OS.  I've authored my fair share of bash scripts to parse data and load it into various data stores (Perl was/is a powerful and pleasant tool - great community; I haven't used it in years).  

I say all this because for some reason, despite my writing a few 10-15 line functions, I don't have a "how to own fishshell" and less the other way around, if you know what I mean.  I have no intentions of going anywhere because I suspect there is reward in eventually "getting it" with fishshell. 

Examples of  my believing "this is more complicated than you think" is when I see error messages, innocuous as they may be, come up after running a `grep` command.  An AppleScript message "expecting end of line but got..." that shows up in a wide range of settings - but not always.  What does fishshell have to do with AppleScript? (it's rhetorical, but I'm curious).  Other things, like having to hit `<ctrl-c>` to get my prompt to show up again after quitting nvim among others.

This is not a rant (no "duck typing please" :)). I truly appreciate and use fishshell all the time.  Healthy relationships can handle a frank conversation.  Nothing is perfect.  The question is, how to make it easier to keep improving the capacity for everyone.  And to be honest, I spend way too much time trying to figure out stuff that is presented in a way that is more complex than need be, or just omits the obvious, but critical level-setting context.

Your explanation was really helpful.  I suspect that others with something that "looks like" what I experienced will have a little more capacity to fish, and not just eat fish :))  My reaction to your feedback is my thinking through "I read the docs", "I have spent time looking at the code from others" but still was completely side-swiped by not knowing anything about `bind` - not even to ask!  Back to the analogy earlier about learning how to use a word processor; it's much easier if you know to ask "how do I copy/paste", more difficult otherwise.  The next time I visit the docs...

Actually, I just spent time revisiting the docs. On the detailed documentation page ([link](http://fishshell.com/docs/current/index.html)), the `bind` sequence shows up 37 times!... how could I have missed it?...  Actually, it turns out that there is one or two lines at most, and as footnotes (essentially).  Prior to today, I read those as "here is how fishshell bridges what you know and enjoy about bash etc. to your experience with fishshell".  A big theme, and understandably so. So, I read "here is a binding..." as a fix, not a feature.  

This brings me to taking a step back.  I wonder if there is value in: "welcome to a brand new world in what you can accomplish (not just how to accomplish it, functions, lists and other syntax-related howtos)".  I can see how, if memory serves, the docs I'm looking at now have been updated.  The "unix philosophy" was a big plus; bash played its part:  working with a filesystem.  I wonder if it's time for fishshell to "own" its share in the pipeline of a users workflow.  Instead of being an alternative, own the task.  Draft the docs accordingly and critically *from scratch*.  To the degree it's practical, maniacally separate the "from bash" perspective. Substitute it with what a user needs to accomplish given the task fishshell owns. Aside, I'm using the word "own" to mean "is the best at", and likely in context of a successful "unix philosophy" (very much in line with the "design philosophy" section).

Is it time to see fishshell in this photo:
![IMG_4342](https://user-images.githubusercontent.com/11245417/102694352-6c4dec00-41ee-11eb-8501-3ab1a0fa0ab9.jpg)

I also gather the docs have gone through a significant change (for instance, I could not find the docs with the blue background). They updates look great.  In past experience I would click on "learn more" to be taken back to where I started... What I have noticed in this latest version is a similar issue.  There are remnants of that same experience in the latest docs.  It's a bizarre experience  because it seems clear but not obvious for some reason.  At the bottom of `tutorial` the "learn more" takes me back to where I started... echos of past frustrations.
> ### Ready for more?
> If you want to learn more about fish, there is lots of detailed documentation, the official gitter channel, an official mailing list, and the github page.

Only just now with this effort to understand "how to fish", for the first time in the many times I checkout out the docs, I failed to see how the links at the top left was where to find more.  I just saw this for the first time:
> ### Other help pages
> Introduction
Commands
Design
Tutorial
Frequently asked questions
License

If in fact the "blue" version of the docs are deprecated, this new version may have mitigated, but I suspect, not eliminated, my prior misunderstanding.

At any rate, first and foremost, thank you to all that build and maintain fishshell.  A critical piece of my productivity workflow.  

Secondly, I think the time it's taking me reach "I own this", is my not fully understanding what fishshell owns.  This an overstatement, but there is truth, I'm stuck thinking from the perspective "bridge, fix etc..".  

In a nutshell, Own it!! Until fishshell does... :))

\- E
